en/lessons/advanced/otp-distribution.md

```diff
@@ -0,0 +1,430 @@
+  doctest Chat
+
+  @tag :distributed
+  test "send_message" do
+    assert Chat.send_message(:moebi@localhost, "hi") == :ok
+  end
+end
+```
+
+And we'll add some conditional logic to our test helper to exclude tests with such tags if the tests are _not_ running on a named node.
+
+```elixir
+exclude =
+  if Node.alive?, do: [], else: [distributed: true]
+
+ExUnit.start(exclude: exclude)
+```
+
+We check to see if the node is alive, i.e. if the node is part of a distributed system with [`Node.alive?`](https://hexdocs.pm/elixir/Node.html#alive?/0). If not, we can tell `ExUnit` to skip any tests with the `distributed: true` tag. Otherwise, we will tell it not to exclude any tests.
+
+Now, if we run plain old `mix test`, we'll see:
+
+```bash
+mix test
+Excluding tags: [distributed: true]
+
+Finished in 0.02 seconds
+1 test, 0 failures, 1 excluded
+```
+
+And if we want to run our distributed tests, we simply need to go through the steps outlined in the previous section: run the `moebi@localhost` node _and_ run the tests in a named node via `iex`.
+
+Let's take a look at our other testing approach--configuring the application to behave differently in different environments.
+
+### Environment-Specific Applicaton Configuration
+
+The part of our code that tells `Task.Supervisor` to start a supervised task on a remote node is here:
+
+```elixir
+# app/chat.ex
+def spawn_task(module, fun, recipient, args) do
+  Task.Supervisor.async(remote_supervisor(recipient), module, fun, args)
+  |> Task.await
+end
+
+defp remote_supervisor(recipient) do
+  {Chat.TaskSupervisor, recipient}
+end
+```
+
+`Task.Supervisor.async` takes in a first argument of the supervisor we want to use. If we pass in a tuple of `{SupervisorName, location}`, it will start up the given supervisor on the given remote node. However, if we pass `Task.Supervisor` a first argument of a supervisor name along, it will use that supervisor to supervise the task locally.
+
+Let's make the `remote_supervisor` function configurable based on environment. In the development environment, it will return `{Chat.TaskSupervisor, recipient}` and in the test environment it will return `Chat.TaskSupervisor`.
+
+We'll do this via application variables.
+
+Create a file, `config/dev.exs`, and add:
+
+```elixir
+# config/dev.exs
+use Mix.Config
+config :chat, remote_supervisor: fn(recipient) -> {Chat.TaskSupervisor, recipient} end
+```
+
+Create a file, `config/test.exs` and add:
+
+```elixir
+# config/test.exs
+use Mix.Config
+config :chat, remote_supervisor: fn(_recipient) -> Chat.TaskSupervisor end
+```
+
+Remember to uncomment this line in `config/config.exs`:
+
+```elixir
+import_config "#{Mix.env()}.exs"
+```
+
+Lastly, we'll update our `Chat.remote_supervisor` function to look up and use the function stored in our new application variable:
+
+```elixir
+# lib/chat.ex
+defp remote_supervisor(recipient) do
+  Application.get_env(:chat, :remote_supervisor).(recipient)
+end
+```
```

Comment:
We typically wrap up our lessons with some small comment other than a code snippet, what do you think about adding something here to summarize what we've covered?